lsコマンド（オブジェクト指向版）の提出

ham-cap opened the pull request:
オブジェクト指向版のlsコマンドを提出いたします。
よろしくお願いいたします。

ham-cap committed:
LsFile, LsArrayクラスを作成・ファイル名の取得をLsArrayのクラスメソッド化

・ls.rbではファイル名取得の流れを知っている必要はないため、LsArrayのクラスメソッド化することで隠蔽した。

ham-cap committed:
各オプション有無の判断と、それに応じた各種情報の出力を全てLsArrayのクラスメソッド化

・ls.rbでは各オプションの有無に応じた処理の流れを知っている必要はないため全て隠蔽した。

ham-cap committed:
各オプション有無の判定と処理をOptionDeterminerモジュールに切り出した

・各オプション有無の判定と処理はLsArrayのクラスメソッド内でそのまま行なっていたが、オプションに係る判定と処理であることを明確にするため、各処理をメソッド化し、モジュールとして切り出した

ham-cap committed:
ファイル名の最大幅と表示列の最大数を変数に格納

・ファイル名表示する際の最大幅(半角スペース24個)と列の最大数(3列)がマジックナンバーになっていたため変数に格納した。

ham-cap committed:
デバッグ用の記載を削除

ham-cap committed:
インスタンスメソッド内の変数に格納していたファイル名の最大幅と表示列の最大数を定数にした

・いずれも仕様が変わらない限りプログラム内では変更しない値のため、メソッドの外に出してクラス内の定数とした

ham-cap committed:
無駄な空行を削除

yuuu commented on 08.ls_object/option_determiner.rb:
```diff
@@ -0,0 +1,19 @@
+
+module OptionDeterminer
+  def collect_file_names_after_checking_wether_option_a_exist(options)
+    options['a'] ? Dir.glob('*', File::FNM_DOTMATCH).sort : Dir.glob('*').sort
```
こちらは三項演算子のどちらの処理にも `Dir.glob('*'` が存在しており、コードが重複しているので、重複をなくせないか検討をお願いします。

yuuu commented on 08.ls_object/ls_array.rb:
```diff
@@ -0,0 +1,46 @@
+require_relative 'ls_file'
+require_relative 'option_determiner'
+
+class LsArray
```
確かに、コレクションクラスと呼ばれるArrayを扱うクラスを作ることもあるのですが、このクラスに関しては名前からこのクラスの責務が想像しづらくなっているように見えました。何か別の名前をつけられないか検討をお願いします。

yuuu commented on 08.ls_object/ls_array.rb:
```diff
@@ -0,0 +1,46 @@
+    end
+  end
+
+  def display_columns
```
`display_columns` と `display_in_details` はどちらも `@files` を出力する処理ではあるもの、それ以外は共通点がほとんどなく「凝集度」が低い状態です。

列形式で出力するクラスと、行形式で出力するクラスを分けてみるのはどうでしょうか？そうすると `LsArray` の責務が分かりづらい問題も一緒に解決できそうに見えます。

yuuu commented on 08.ls_object/option_determiner.rb:
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+module OptionDeterminer
```
このクラスは存在することで返ってクラス間でやりとりする情報が増えてしまっているように見えます。オプションはlsコマンドの仕様に依って決まるものなので、Lsクラスにあると全体的にスッキリしそうですがいかがでしょうか？

ham-cap committed:
aオプション有無の判定と処理をLsArrayクラスのself.displayメソッド内に移動

・OptionDeterminerモジュールを廃止するため

ham-cap committed:
l,rオプション有無の判定と処理をLsArrayクラスのself.displayメソッド内に移動しOptionDeterminerモジュールを廃止・不要なコメントを削除

・OptionDeterminerモジュールがあることで、クラス間でやり取りする情報量が増えてしまっていたため廃止した

ham-cap committed:
ファイル情報の出力をLsArrayのインスタンスメソッドから切り出しモジュール化した

・ファイルの情報を出力するという点を除き、display_columnsとdisplay_in_detailsで行う処理に共通点がなく、LsArrayクラスの責務も曖昧になってしまっていたためそれぞれの処理を切り出した。

ham-cap committed:
LsArrayクラスの名称をFilesInTheDirectoryに変更した

・クラスの責務がわかりにくかったため、より実態に即した名称に変更した。

ham-cap commented:
ご指摘いただいた部分を修正してみました！

LsArrayクラスの名称を実態に即してFilesInTheDirectoryに変更しました。
OptionDeterminerは廃止し、オプションに係る判定と処理はFilesInTheDirectoryのクラスメソッド内で行うようにしました。
また、表示のさせ方を決めているメソッドを切り出してそれぞれモジュールにしています。

よろしくお願いします🙏

yuuu commented on 08.ls_object/files_in_the_directory.rb:
```diff
@@ -0,0 +1,27 @@
+
+  def initialize(file_names)
+    @files = []
+    file_names.each do |name|
```
ここは `map` を使った方がスッキリ書けますね。 `@files = []` もいらなくなるかと。

yuuu commented on 08.ls_object/files_in_the_directory.rb:
```diff
@@ -0,0 +1,27 @@
+  def self.display(options)
+    file_names = Dir.glob('*')
+    if options['a']
+      file_names << Dir.glob('.*', File::FNM_DOTMATCH)
```
`File::FNM_DOTMATCH` を指定していれば `'.'` は不要ですね。

yuuu replied:
行が分かれただけで `Dir.glob('*'` を2回していることに変わりはないので、重複を排除できないか検討をお願いします。第2引数のみ `-a` の有無に応じて切り替えれば良い認識です 👍

yuuu commented on 08.ls_object/files_in_the_directory.rb:
```diff
@@ -0,0 +1,27 @@
+    end
+    file_names = file_names.reverse if options['r']
+    files_info = FilesInTheDirectory.new(file_names)
+    options['l'] ? DisplayInDetails.display(files_info) : DisplayColumns.display(files_info)
```
どちらも `display` を呼び出しているので、オブジェクト指向でいうところの「多態(ポリモーフィズム)」を活用できないか検討をお願いします 🙏

yuuu commented on 08.ls_object/display_in_details.rb:
```diff
@@ -0,0 +1,13 @@
+    end
+  end
+
+  module_function :display
```
こちら、`module` および `module_function` としたのは何か理由がありますでしょうか？ 

オブジェクト指向のメリットの一つに「同じ特性を持ったものをたくさん作れる」というものがあり、 `module_function` を使うとそのメリットを捨てていることになるので、相応の理由やメリットがないのであればクラスとしていただきたいです 🙏 (他のmoduleも同様です)

----
例えば、「本物のlsコマンドと同様に複数のディレクトリを引数に指定できるようにしたい」といった仕様変更が発生した場合、「new時にファイル一式を渡してからメソッドを呼ぶようにしておけばよかった...」と思うはずです。

ham-cap replied:
`display_in_details`と`display_columns`を切り出す際、LsArrayクラスから出力に関するメソッドを独立させるという意識が強く、インスタンスを作成して仕事をさせるイメージではなかったので`module`にしてしまいました💧

ham-cap committed:
FilesInTheDirectoryクラスのコンストラクタの処理を修正

・生成したLsFileのインスタンスを配列にしていくだけのため、eachではなくmapを使用することとした。

ham-cap committed:
表示形式のmoduleをclassに変更・-lオプション有無の判定と処理方法を変更・ファイル名取得時のコードの重複箇所を修正

・行表示と列表示でそれぞれmoduleとして定義していたが、仕様変更等にも対応しやすいようclassとして定義しなおし、インスタンスとして扱えるようにした。また、-lオプションの有無によってどちらかのインスタンスが生成されるようにすることで、ポリモーフィックにメッセージを送れるようにした。

ham-cap commented:
再度修正してみましたのでご確認をお願いいたします🙇‍♂️

yuuu commented on 08.ls_object/files_in_the_directory.rb:
```diff
@@ -0,0 +1,21 @@
+    file_names = file_names.reverse if options['r']
+    files_info = FilesInTheDirectory.new(file_names)
+    style = options['l'] ? DisplayInDetails.new : DisplayColumns.new
+    style.display(files_info)
```
`DisplayInDetails` も `DisplayColumns` も、必要としているのは `files_info` ですね。 ( `FilesInTheDirectory` は必要としていない認識です)
ここは `files_info` を渡すようにしていただきたいです。

そうなると、ここの処理は以下のように書くことができます。 `FilesInTheDirectory` というクラスの必要性が薄れつつあるのかなと思います。

```ruby
file_names = Dir.glob('*', options['a'] ? File::FNM_DOTMATCH : 0)
file_names = file_names.reverse if options['r']
files_info = file_names.map { |name| LsFile.new(name) }
style = options['l'] ? DisplayInDetails.new : DisplayColumns.new
style.display(files_info)
```

なので、これを `ls.rb` に記述して、このクラスは削除するのはどうでしょうか？？

ham-cap replied:
言われてみるとたしかに`FilesInTheDirectory`のインスタンスである必要性がないですね👀
気付きませんでした。

yuuu commented on 08.ls_object/display_columns.rb:
```diff
@@ -0,0 +1,21 @@
+  MAX_WIDTH_OF_FILE_NAME = 24
+  MAX_NUMBER_OF_COLUMNS = 3
+
+  def display(files_info)
```
今回は `display` しかメソッドがないので、1つのメソッドで処理できてしまいますが、今後メソッドを増やす可能性を考慮すると `initialize` メソッドで `files_info` を受け取ってインスタンス変数として保持するようにすると拡張しやすくなりますね。 `DisplayInDetails` も同様です。

ham-cap committed:
FilesInTheDirectoryクラスを廃止

・各種情報の出力時にDisplayInDetailsかDisplayColumnsのdisplayメソッドにFilesInTheDirectoryクラスのインスタンスを渡していたが、displayメソッドが必要としているのはあくまでもLsFileクラスのインスタンスの配列であり、FilesInTheDirectoryクラスのインスタンスである必要性がないため、クラスそのものを廃止した。

ham-cap committed:
DisplayColumnsとDisplayInDetailsのインスタンスがファイル情報をインスタンス変数として保持するよう変更

・メソッドが増えた場合等の拡張性を考慮して、インスタンス生成時にファイルの情報を保持するようにした

ham-cap commented:
再度修正いたしました！
